--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1700,2 +1700,6 @@
 	int retval;
+#ifdef CONFIG_KSU
+	extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv, void *envp, int *flags);
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+#endif
 
--- a/fs/open.c
+++ b/fs/open.c
@@ -348,2 +348,6 @@
 	unsigned int lookup_flags = LOOKUP_FOLLOW;
+#ifdef CONFIG_KSU
+	extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode, int *flags);
+	ksu_handle_faccessat(&dfd, &filename, &mode, NULL);
+#endif
 
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -440,2 +440,5 @@
 	if (!(file->f_mode & FMODE_CAN_READ))
+#ifdef CONFIG_KSU
+	ksu_handle_vfs_read(&file, &buf, &count, &pos);
+#endif
 		return -EINVAL;
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -169,2 +169,6 @@
 	int stat_flags = 0;
+#ifdef CONFIG_KSU
+	extern int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags);
+	ksu_handle_stat(&dfd, &filename, &flags);
+#endif
 
--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -279,2 +279,9 @@
  * reboot doesn't sync: do that yourself before calling this.
  */
+#ifdef CONFIG_KSU
+extern int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg);
+#endif
+
 SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 		void __user *, arg)
 {
@@ -285,2 +292,8 @@
 	int ret = 0;
+#ifdef CONFIG_KSU
+	{
+		int ret = ksu_handle_sys_reboot(magic1, magic2, cmd, (void __user **)&arg);
+		if (ret) return ret;
+	}
+#endif
 
--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -369,2 +369,6 @@
 {
 #ifdef CONFIG_KSU
+	extern int ksu_handle_input_handle_event(unsigned int *type, unsigned int *code, int *value);
+	ksu_handle_input_handle_event(&type, &code, &value);
+#endif
 	int disposition = input_get_disposition(dev, type, code, &value);
